subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingClasspathFileTransformer.java

```diff
@@ -76,6 +76,12 @@ private static ClasspathFileHasher createFileHasherWithConfig(HashCode configHas
             Hasher hasher = Hashing.newHasher();
             hasher.putHash(configHash);
             hasher.putHash(fileHasher.hashOf(sourceSnapshot));
+            if (sourceSnapshot.getType() == FileType.Directory) {
+                // Prior to 8.7 we were combining instrumented classes from both directories and JARs into JARs.
+                // Now we store instrumented directories as directories, so we invalidate these.
+                // However, transformed JARs should be left intact.
+                hasher.putBoolean(true);
```

Comment:
💅  Since we now use artifact transforms/execution engine for most of instrumentation, where caching is bound to Gradle version, I wonder if we can simplify this to just `hasher.putBoolean(sourceSnapshot.getType() == FileType.Directory);`  without any `if`

Reply:
You can also leave that.

But that is a good reminder that I should do the cleanup soon of this custom instrumentation.